File: 06_Algorithm/leetcode_054.go
/**
 * File              : leetcode_054.go
 * Author            : Andy
 * Date              : 2022.10.01
 * Last Modified Date: 2022.10.01
 * Last Modified By  : Andy
 */

package convertBST

func convertBST(root *TreeNode) *TreeNode {
	total := 0
	var dfs func(*TreeNode, *int) *TreeNode
	dfs = func(node *TreeNode, total *int) *TreeNode {
		if node == nil {
			return nil
		}

		dfs(node.Right, total)
		*total += node.val
		node.Val = *total
		dfs(node.Left, total)
		return node
	}
	return dfs(root, &total)
}

File: Algorithm/leetcode_054.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
# File              : leetcode_054.py
# Author            : Andy
# Date              : 2022.10.01
# Last Modified Date: 2022.10.01
# Last Modified By  : Andy


# 尝试优先，后序遍历，先找到最右侧的节点
class Solution:

    def convertBST(self, root):
        total = 0

        def dfs(node):
            nonlocal total

            if node is None:
                return

            dfs(node.right)
            total += node.val
            node.val = total
            dfs(node.left)
            return node

    return dfs(root)
